TL/MLX5: adding mcast helper functions

MamziB opened the pull request:
TL/MLX5: adding mcast helper functions

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,199 @@
+
+    if (0 != rdma_resolve_addr(ctx->id, (struct sockaddr *)&ip_oib_addr,
+                (struct sockaddr *) &dst_addr, 1000)) {
+        tl_error(lib, "failed to resolve rdma addr, errno %d", errno);
```
not an error

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,199 @@
+        goto error;
+    }
+
+    ib           = strdup(ctx->devname);
```
alignment

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,199 @@
+        goto error;
+    }
+
+    if (port_attr.max_mtu == IBV_MTU_256)
+        max_mtu = 256;
+    if (port_attr.max_mtu == IBV_MTU_512)
+        max_mtu = 512;
+    if (port_attr.max_mtu == IBV_MTU_1024)
+        max_mtu = 1024;
+    if (port_attr.max_mtu == IBV_MTU_2048)
+        max_mtu = 2048;
+    if (port_attr.max_mtu == IBV_MTU_4096)
+        max_mtu = 4096;
```
pls use lookup array

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,199 @@
+                device_attr.max_cq, device_attr.max_cqe);
+
+    ctx->max_qp_wr = device_attr.max_qp_wr;
+    {
```
braces are not needed

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,199 @@
+    ctx->rcache = NULL;
+    if (UCC_OK != ucc_tl_mlx5_mcast_setup_rcache(ctx)) {
+        tl_error(lib, "failed to setup rcache");
+        status = UCC_ERR_NO_RESOURCE;
```
return actual status

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,199 @@
+        goto error;
+    }
+
+    tl_debug(ctx->lib, "tl mcast setup complete: ctx %p", ctx);
```
"tl mcast" is not needed, it's added by tl_debug.
"setup complete" is ambiguous, "context setup complete" may be better

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_team.c:
```diff
@@ -18,3 +19,94 @@ ucc_status_t ucc_tl_mlx5_mcast_team_init(ucc_base_context_t           *base_cont
+    buf_size  = comm->ctx->mtu;
+
+    // Comm receiving buffers.
+    posix_memalign((void**)&comm->call_rwr, page_size, sizeof(struct ibv_recv_wr) *
```
check return code

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_team.c:
```diff
@@ -18,3 +19,94 @@ ucc_status_t ucc_tl_mlx5_mcast_team_init(ucc_base_context_t           *base_cont
+    comm->tx                           = 0;
+    comm->n_prep_reliable              = 0;
+    comm->n_mcast_reliable             = 0;
+    comm->reliable_in_progress         = false;
```
what is "false"? is it C standard bool? We use integers instead

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,199 @@
+    }
+
+    ib           = strdup(ctx->devname);
+    ucs_string_split(ib, ":", 2, &ib_name, &port);
```
don't use ucs functions

Sergei-Lebedev reviewed (commented):
@MamziB please fix CI

MamziB commented:
@Sergei-Lebedev Thanks for the comments, I appreciate your time. I incorporated your comments, and updated the commit.
 Please note that HPC_SDK fails somewhere out of this PR scope:
````
"/__w/ucc/ucc/src/coll_patterns/double_binary_tree.h", line 55: error #68: integer conversion resulted in a change of sign [integer_sign_change]
          return -1;
```
@samnordmann Please kindly take a look at this PR and let me know your thoughts. Your constructive comments are much appreciated.

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+    if (num_devices >= 1 && !strcmp(mcast_ctx_conf->ib_dev_name, "")) {
+        dev          = device_list[0];
+        devname      = (char *)ibv_get_device_name(dev);
+        ctx->devname = ucc_malloc(strlen(devname)+16, "devname");
```
looks like it is not getting freed in mcast cleanup. I added the needed code

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+            goto error;
+        }
+        strncpy(ctx->devname, devname, strlen(devname));
+        strncat(ctx->devname, ":1", 16);
```
yes we are doing it when later in this function when we want to determine MTU

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+        user_provided_ib = 0;
+    } else {
+        /* user has provided the devname now make sure it is valid */
+        for (i = 0; device_list[i]; ++i) {
```
- because we are not using mlx5dv.h in our mcast code.
- same as above - also it gives flexibility to us to choose different devices for multicast and alltoall 
- I removed that check and added``` ib_valid ``` variable.

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+        tl_debug(lib, "failed to get ipoib interface for devname %s", ctx->devname);
+        status = UCC_ERR_NOT_SUPPORTED;
+        if (!user_provided_ib) {
+            ucc_free(ctx->devname);
```
it is just for a case where we do malloc a few lines earlier, so that is what we added here in this function and we should free it

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+        goto error;
+    }
+
+    if (rdma_get_cm_event(ctx->channel, &revent) < 0 ||
```
looks like yes, I added rdma_ack_cm_event() to free the event

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+    }
+
+    rdma_ack_cm_event(revent);
+    ctx->ctx = ctx->id->verbs;
```
i think rdma cm library is responsible for that during rdma_create_id().

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+        goto error;
+    }
+
+    ib = strdup(ctx->devname);
```
first ```str``` input in ucc_string_split() is not const, so we wanna make sure this function does not change ctx->devname

MamziB replied:
@Sergei-Lebedev it is so weird, not sure when strdup() got removed from code. Anyway, thanks for catching it.

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+    ib = strdup(ctx->devname);
+    ucc_string_split(ib, ":", 2, &ib_name, &port);
+    ctx->ib_port = atoi(port);
+
```
yes I freed it

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,219 @@
+    ctx->ib_port = atoi(port);
+
+    /* Determine MTU */
+    if (ibv_query_port(ctx->ctx, ctx->ib_port, &port_attr)) {
```
this function does not return the port attributes that we need.

MamziB commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_helper.c:
```diff
@@ -189,3 +189,345 @@ ucc_status_t ucc_tl_mlx5_probe_ip_over_ib(char* ib_dev, struct
+        }
+    }
+
+    if (comm->grh_mr) {
```
when we create a comm object, we set it all to 0, so over there is getting initialized.

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
-ucc_status_t ucc_tl_mlx5_mcast_context_init(ucc_tl_mlx5_mcast_context_t    *context, /* NOLINT */
-                                            ucc_tl_mlx5_mcast_ctx_params_t *mcast_ctx_conf /* NOLINT */)
+#define UCC_TL_MLX5_MCAST_MAX_MTU_COUNT 5
+int mtu_lookup[UCC_TL_MLX5_MCAST_MAX_MTU_COUNT][2] =
```
pls use formatting similar to https://github.com/openucx/ucc/blob/a2ae31ce7aafe8970d9cbaf63f2a21e24bda2832/src/components/tl/sharp/tl_sharp_coll.c#L15

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+    int                     active_mtu    = 4096;
+    int                     max_mtu       = 4096;
+    ucc_tl_mlx5_mcast_coll_context_t *ctx = NULL;
+
```
remov blank line

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+    lib      = mlx5_ctx->super.super.lib;
+    ctx->lib = lib;
+
+    device_list = ibv_get_device_list(&num_devices);
```
i think this check should be common for TL MLX5, it's not multicast specific

MamziB replied:
I think in the future when mlx5 TL is mature, we must unify all the contexts that we create in mlx5 TL. For now, we can continue with separate ctx. I will add this item in the to-do list for future.

Sergei-Lebedev replied:
Sharing resources might be tricky since mlx5 alltoall need specific features e.g. shared pd. But at least checking ib devices should be easy

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+            goto error;
+        }
+        strncpy(ctx->devname, devname, strlen(devname));
+        strncat(ctx->devname, ":1", 16);
```
it should be smaller than 16 since part of the string is already occupied with devname

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+            }
+        }
+        if (!ib_valid) {
+            tl_debug(lib, "ib device %s not found", mcast_ctx_conf->ib_dev_name);
```
in that case it's warning probably since user provided incorrect config value

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+        if (!ib_valid) {
+            tl_debug(lib, "ib device %s not found", mcast_ctx_conf->ib_dev_name);
+            status = UCC_ERR_NOT_FOUND;
+            goto error;
```
need to call ibv_free_device_list if you exit from function

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+    ibv_free_device_list(device_list);
+    if (UCC_OK != ucc_tl_mlx5_probe_ip_over_ib(ctx->devname, &ip_oib_addr)) {
+        tl_debug(lib, "failed to get ipoib interface for devname %s", ctx->devname);
+        status = UCC_ERR_NOT_SUPPORTED;
```
ues status returned from ucc_tl_mlx5_probe_ip_over_ib

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+
+    ctx->channel = rdma_create_event_channel();
+    if (!ctx->channel) {
+        tl_error(lib, "rdma_create_event_channel failed, errno %d", errno);
```
is it really an error? should we always be able to create even channel if device is up?

MamziB replied:
I can change it to warning then.

Sergei-Lebedev replied:
in general, if the error is not critical we shouldn't print errors or warnings. It won't pass QA with such output.

MamziB replied:
@Sergei-Lebedev Sure, I changed it to tl_debug()

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+    }
+
+
+    for (i = 0; i < UCC_TL_MLX5_MCAST_MAX_MTU_COUNT; i++) {
+        if (mtu_lookup[i][1] == port_attr.max_mtu) {
+            max_mtu = mtu_lookup[i][0];
+            break;
+        }
+    }
```
why do you need for loop? max_mtu is enum just simple max_mtu = mtu_lookup[ptrt_attr.max_mtu]

MamziB replied:
what if in the future the order of enum list changes and we forget to update that mtu_lookup array.  I wanted us to be on the safe side, otherwise I could simply use 2^(7+enum) and get the mtu

Sergei-Lebedev replied:
I don't think rdma core will ever change this enum and break ABI. Also with such change how your approach will work?

MamziB replied:
enum starts from 1. So, for like IBV_MTU_256 == 1, to achieve 256, I could do pow(2, IBV_MTU_256+7) and get 2^8 = 256 out of it. Should I still change this for loop? leave it as is?

Sergei-Lebedev replied:
I don't mind using pow 2 or array lookup or even switch over enum, just avoid duplication of same code.

MamziB replied:
alright, I just merged two for loops into a single one to avoid duplication.

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:
```diff
@@ -10,9 +10,241 @@
+        goto error;
+    }
+
+    tl_debug(ctx->lib, "context setup complete: ctx %p", ctx);
```
multicast context setup complete

Sergei-Lebedev commented on src/components/tl/mlx5/mcast/tl_mlx5_mcast_helper.c:
```diff
@@ -189,3 +189,367 @@ ucc_status_t ucc_tl_mlx5_probe_ip_over_ib(char* ib_dev, struct
+    if (ctx->pkey_index >= port_attr.pkey_tbl_len) {
+        ctx->pkey_index = 0;
+        ibv_query_pkey(ctx->ctx, ctx->ib_port, ctx->pkey_index, &pkey);
+        if (pkey) {
+            tl_debug(ctx->lib, "cannot find default pkey 0x%04x on port %d, using index 0 pkey:0x%04x",
+                       DEF_PKEY, ctx->ib_port, pkey);
+        } else {
+            tl_error(ctx->lib, "cannot find valid PKEY");
+            return UCC_ERR_NO_RESOURCE;
+        }
```
rewrite without else

MamziB commented:
Thank you @Sergei-Lebedev for the latest comments, I have updated the commit. @samnordmann Please kindly let me know if you have further thoughts about the PR otherwise we can finalize it

MamziB commented:
@Sergei-Lebedev build UCC is failing in this PR because of some issues unrelated to it:

````
"/__w/ucc/ucc/src/coll_patterns/double_binary_tree.h", line 139: error #68: integer conversion resulted in a change of sign [integer_sign_change]
````

MamziB commented:
@Sergei-Lebedev This error is not related to this PR. If it was, it should have printed this:
```
failed to move mcast qp to INIT
````

MamziB commented:
@Sergei-Lebedev I just updated the commit. Please let me know if you have further comments. Thanks

Sergei-Lebedev commented:
@MamziB please fix CI failures
